src/ui/MessageMenu/messageMenu.tsx

```diff
@@ -0,0 +1,238 @@
+import { OpenChannel } from '@sendbird/chat/openChannel';
+
+import { SendableMessageType, copyToClipboard, isFailedMessage, isUserMessage } from '../../utils';
+import { classnames, noop } from '../../utils/utils';
+import useSendbirdStateContext from '../../hooks/useSendbirdStateContext';
+import ContextMenu, { MenuItems } from '../ContextMenu';
+import { MenuItem, PrebuildMenuItemPropsType, TriggerIcon, TriggerIconProps } from './items';
+import { useLocalization } from '../../lib/LocalizationContext';
+import { ReplyType } from '../../types';
+import {
+  MenuConditionsParams,
+  showMenuItemCopy,
+  showMenuItemDelete,
+  showMenuItemEdit,
+  showMenuItemOpenInChannel,
+  showMenuItemReply,
+  showMenuItemResend,
+  showMenuItemThread,
+} from '../../utils/menuConditions';
+
+interface RenderTriggerParams extends TriggerIconProps {
+  TriggerIcon: (props: TriggerIconProps) => ReactElement;
+}
+type PrebuildItemsType = { [key: string]: (props: PrebuildMenuItemPropsType) => ReactElement };
+type RenderMenuItemsParams = {
+  close: () => void;
+  prebuildItems: PrebuildItemsType;
+};
+
+export interface MessageMenuProps {
+  className?: string;
+  message: SendableMessageType;
+  channel: GroupChannel | OpenChannel | null;
+  isByMe?: boolean;
+  replyType?: ReplyType;
+  renderTrigger?: (params: RenderTriggerParams) => ReactElement;
+  renderMenuItems?: (params: RenderMenuItemsParams) => ReactElement;
+
+  disableDeleteMessage?: boolean;
+  showEdit?: (bool: boolean) => void;
+  showRemove?: (bool: boolean) => void;
+  deleteMessage?: (message: SendableMessageType) => void;
+  resendMessage?: (message: SendableMessageType) => void;
+  setQuoteMessage?: (message: SendableMessageType) => void;
+  setSupposedHover?: (bool: boolean) => void;
+  onReplyInThread?: (props: { message: SendableMessageType }) => void;
+  onMoveToParentMessage?: () => void;
+}
+export const MessageMenu = ({
+  className,
+  message,
+  channel,
+  isByMe,
+  replyType,
+  renderTrigger = TriggerIcon,
+  renderMenuItems,
+
+  disableDeleteMessage = null,
+  showEdit = noop,
+  showRemove = noop,
+  deleteMessage,
+  resendMessage,
+  setQuoteMessage,
+  setSupposedHover,
+  onReplyInThread,
+  onMoveToParentMessage,
+}: MessageMenuProps) => {
+  const { stringSet } = useLocalization();
+  const { config } = useSendbirdStateContext();
+  const { isOnline } = config;
+
+  const triggerRef = useRef(null);
+  const containerRef = useRef(null);
+
+  const params: MenuConditionsParams = {
+    message,
+    channel,
+    isByMe,
+    replyType,
+    onReplyInThread,
+    onMoveToParentMessage,
+  };
+
+  // # Menu Items
+  const Copy = (props: PrebuildMenuItemPropsType) => (
+    <MenuItem
+      {...props}
+      onClick={(e) => {
+        if (isUserMessage(message)) copyToClipboard(message?.message);
+        props?.onClick?.(e);
+      }}
+    >
+      {props?.children ?? stringSet.MESSAGE_MENU__COPY}
+    </MenuItem>
+  );
```

Comment:
Do not create and use components during rendering. Always define components at the top level.
https://react.dev/learn/your-first-component#nesting-and-organizing-components

Reply:
You can pass `onClose`, `onClickCopy`, `onClickDelete`, ... through context.